titus-common-ext/kube/src/main/java/com/netflix/titus/ext/kube/clustermembership/connector/transport/fabric8io/Fabric8IOKubeMembershipExecutor.java

```diff
@@ -0,0 +1,155 @@
+                    .withLabels(CollectionsExt.copyAndRemove(localMemberRevision.getCurrent().getLabels(), "resourceVersion"))
+                    .build();
+            ClusterMembershipRevision<ClusterMember> noVersionRevision = localMemberRevision.toBuilder().withCurrent(currentWithoutVersion).build();
+
+            FClusterMembership crd = Fabric8IOModelConverters.toCrdFClusterMembership(noVersionRevision);
+            FClusterMembership storedCrd = membershipClient.create(crd);
+
+            return applyNewRevisionNumber(
+                    Fabric8IOModelConverters.toClusterMembershipRevision(storedCrd.getSpec()),
+                    storedCrd.getMetadata()
+            );
+        }).subscribeOn(scheduler);
+    }
+
+    @Override
+    public Mono<ClusterMembershipRevision<ClusterMember>> updateLocal(ClusterMembershipRevision<ClusterMember> localMemberRevision) {
+        return Mono.fromCallable(() -> {
+            FClusterMembership crd = Fabric8IOModelConverters.toCrdFClusterMembership(localMemberRevision);
+            FClusterMembership storedCrd = membershipClient.withName(localMemberRevision.getCurrent().getMemberId()).replace(crd);
+            return applyNewRevisionNumber(
+                    Fabric8IOModelConverters.toClusterMembershipRevision(storedCrd.getSpec()),
+                    storedCrd.getMetadata()
+            );
+        }).subscribeOn(scheduler);
+    }
+
+    @Override
+    public Mono<Void> removeMember(String memberId) {
+        return Mono.<Void>fromRunnable(() -> {
+            membershipClient.withName(memberId).delete();
+        }).subscribeOn(scheduler);
+    }
+
+    @Override
+    public Flux<ClusterMembershipEvent> watchMembershipEvents() {
+        return Flux.create(sink -> {
+            membershipClient.watch(
+                    new Watcher<FClusterMembership>() {
+                        @Override
+                        public void eventReceived(Action action, FClusterMembership resource) {
+                            ClusterMembershipRevision<ClusterMember> revision = applyNewRevisionNumber(
+                                    Fabric8IOModelConverters.toClusterMembershipRevision(resource.getSpec()),
+                                    resource.getMetadata()
+                            );
+                            switch (action) {
+                                case ADDED:
+                                    sink.next(ClusterMembershipEvent.memberAddedEvent(revision));
+                                    break;
+                                case MODIFIED:
+                                    sink.next(ClusterMembershipEvent.memberUpdatedEvent(revision));
+                                    break;
+                                case DELETED:
+                                    sink.next(ClusterMembershipEvent.memberRemovedEvent(revision));
+                                    break;
+                                case ERROR:
```

Comment:
do you know if the fabric8 client guarantees the watch stream will be closed after all `ERROR` actions?

Reply:
No, it will not be closed. As I understand this is potentially recoverable error. But the callback does not provide any information on the nature of the error, so there is nothing to be logged here.

Here you have code that handles it:
```java
  protected boolean onStatus(Status status) {
    // The resource version no longer exists - this has to be handled by the caller.
    if (status.getCode() == HTTP_GONE) {
      close(new WatcherException(status.getMessage(), new KubernetesClientException(status)));
      return true;
    }

    eventReceived(Action.ERROR, null);
    logger.error("Error received: {}", status);
    return false;
  }
```

`close` is call if the error is not recoverable, which happens only if the resource itself is gone.